File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-3.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if len(s) < 2:
            return len(s)

        sol = 1

        substrings = defaultdict(bool)
        substrings[s[0]] = True

        left = 0
        right = 1

        current = 1

        # x y z z y
        while right < len(s):
            if substrings[s[right]] == False:
                substrings[s[right]] = True
                right += 1
                current += 1
                sol = max(sol, current)
            else:
                if s[left] != s[right]:
                    substrings[s[left]] = False
                    current -= 1  
                else:
                    right += 1
                left += 1        

        return sol

File: Data Structures & Algorithms/spiral-matrix/submission-3.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        vector<int> sol;

        int m = matrix.size();
        int n = matrix[0].size();

        int top = 0;
        int bottom = m - 1;

        int left = 0;
        int right = n - 1;

        while (top <= bottom and left <= right) {
            for (int c = left; c <= right; c++) {
                sol.push_back(matrix[top][c]);
            }

            for (int r = top + 1; r < bottom; r++) {
                sol.push_back(matrix[r][right]);
            }

            if (top < bottom) {
                for (int c = right; c >= left; c--) {
                    sol.push_back(matrix[bottom][c]);
                }
            }

            if (left < right) {
                for (int r = bottom - 1; r > top; r--) {
                    sol.push_back(matrix[r][left]);
                }
            }

            top++;
            bottom--;
            left++;
            right--;
        }

        return sol;
    }
};
